### ISSUE TITLE
Various DOS Program Bugs under emu2

### ISSUE THREAD

**Issue by johnsonjh** [dmsc/emu2#39]
* ~[**IBM PKZIP/PKUNZIP 2.06**](http://citadel.ringoflightning.net/pkzip206.rar) works correctly, but without displaying any output.~ (*Fixed*)
* [**IBM FileCommand II**](https://winworldpc.com/product/filecommand/ii) paints a blue screen and does not continue.
* ~[**IBM MYMENU**](https://winworldpc.com/product/ibm-professional-editor/ii) (part of *IBM Professional Editor II V1.35*) does not start.~ (*Fixed*)
* [**IBM Professional Editor II V1.35**](https://winworldpc.com/product/ibm-professional-editor/ii) (*after un-EXEPACK'ing* or *LOADFIX*) works fine for all editing functions, but locks up when attempting to display directory listings. (*[Update](https://github.com/dmsc/emu2/issues/39#issuecomment-1089705245)*)
* [Fifth Gen. Systems **Direct Access 5.19**](https://winworldpc.com/product/direct-access/5x) does not start once uncompressed or installed. 
  * When using the **DA Decompression Utility** (`DA_DCMP.EXE`, i.e. `DA_DCMP.EXE LOGO_TXT.DA5`), filenames that are less than eight characters are incorrectly padded with spaces, for example `LOGO.TXT` becomes `LOGO.TXT    `.
* While of limited utility (as later MSC releases require a 386), **Microsoft Programmer's Workbench IDE for DOS** (`PWB.EXE`) displays a spurious error starting from the MSC7 `BIN` directory, (which doesn't happen under DOS): 
* ![D60CD6D1-DC42-4F11-80E0-E6C4A20D603B](https://user-images.githubusercontent.com/61629094/161829767-f5cedbd9-15bf-4f0e-810d-d845b6ceeed2.jpeg)

I haven't looked into causes myself, but perhaps I can this weekend. 

I can also split these into separate issues, if you prefer.

**Comment by johnsonjh** [comment]
And one more, using **VEDIT PLUS v4.12** (*12/07/94*) under emu2, the first directory entry appears twice in listings, and the `..` entry does not appear, so you cannot navigate back into the parent directory you started from: ![Screenshot from 2022-04-05 15-24-38](https://user-images.githubusercontent.com/61629094/161834063-b77da1a4-4c7f-43e1-aeb5-9e5bb43c705f.png)

**Comment by dmsc** [comment]
Hi!

Yes, I would prefer separate issues, as it is easier to track progress.

> * [**IBM PKZIP/PKUNZIP 2.06**](http://citadel.ringoflightning.net/pkzip206.rar) works correctly, but without displaying any output.

I fixes this in the last commits, a2d052926f1208344a57f50d20f215a5c693f39d fixes the PKZIP screen output, d17699c9d6111d021c9dd868796e7355df749a06 fixes PKCFG program.

> * [**IBM FileCommand II**](https://winworldpc.com/product/filecommand/ii) paints a blue screen and does not continue.

This programs loops reading the CGA registers, counting the horizontal retraces, currently emu2 simply fakes the values, so it won't work without some complex video emulation.

Have Fun!

**Comment by dmsc** [comment]
Hi!

> * [**IBM MYMENU**](https://winworldpc.com/product/ibm-professional-editor/ii) (part of _IBM Professional Editor II V1.35_) does not start.

This program expects the timer interrupt to trigger before starting. Fixed in e9c17a9df4f2fd70abbd8c2cbcc4f16dc7e7b66c

> * [**IBM Professional Editor II V1.35**](https://winworldpc.com/product/ibm-professional-editor/ii) (_after un-EXEPACK'ing_ or _LOADFIX_) works fine for all editing functions, but locks up when attempting to display directory listings.

The program has a bug, it does not support more than 32767 clusters free in the disk, and locks-up if any higher number is available. This patch make it work:
```
diff --git a/src/dos.c b/src/dos.c
index 58ae440..0602418 100644
--- a/src/dos.c
+++ b/src/dos.c
@@ -1410,7 +1410,7 @@ void int21()
         break;
     case 0x36: // get free space
         cpuSetAX(32);     // 16k clusters
-        cpuSetBX(0xFFFF); // all free, 1GB
+        cpuSetBX(0x7FFF); // all free, 512MB
         cpuSetCX(512);    // 512 bytes/sector
         cpuSetDX(0xFFFF); // total 1GB
         break;
```

Perhaps I will apply this, I don't know if other programs rely on the disk free information.

Have Fun!

**Comment by johnsonjh** [comment]
@dmsc 

Wow, great work - you're fast and almost down my list!  I was going to split it out into different issues, but I'll wait.

Perhaps for the Professional Editor patch you mention above, a better thing to do would be an option similar to `EMU2_LOWMEM` ... maybe `EMU2_SMALLDISK`?

**Comment by johnsonjh** [comment]
I'll move the remaining now.

**Comment by johnsonjh** [comment]
*Split into separate issues:*
1. https://github.com/dmsc/emu2/issues/44
2. https://github.com/dmsc/emu2/issues/43
3. https://github.com/dmsc/emu2/issues/42
4. https://github.com/dmsc/emu2/issues/41
5. https://github.com/dmsc/emu2/issues/40
